src/NHibernate/Engine/ActionQueue.cs

```diff
@@ -565,36 +579,47 @@ private int GetBatchNumber(EntityInsertAction action, string entityName)
 				// so specify that this entity is with the latest batch.
 				// doing the batch number before adding the name to the list is
 				// a faster way to get an accurate number.
-
 				batchNumber = _actionBatches.Count;
 				_latestBatches[entityName] = batchNumber;
 				return batchNumber;
 			}
 
-			private bool IsProcessedAfterAllAssociatedEntities(EntityInsertAction action, int latestBatchNumberForType)
+			private bool RequireNewBatch(EntityInsertAction action, int latestBatchNumberForType)
 			{
+				// This method assumes the original action list is already sorted in order to respect dependencies.
 				var propertyValues = action.State;
-				var propertyTypes = action.Persister.ClassMetadata.PropertyTypes;
+				var propertyTypes = action.Persister.EntityMetamodel?.PropertyTypes;
```

Comment:
`ClassMetadata` current NHibernate implementation is reaching the underlying `EntityMetamodel.PropertyTypes`. As `ClassMetadata` seems "more optional" than `EntityMetamodel`, better use `EntityMetamodel` directly. (`CustomPersister` on NHibernate side does not provide any of them, but on Hibernate side, it does provide `EntityMetamodel` while still not providing `ClassMetadata`.)